whylog/config/abstract_config.py

```diff
@@ -0,0 +1,135 @@
+        pass
+
+    @abstractmethod
+    def _load_rules(self):
+        pass
+
+    @abstractmethod
+    def _load_log_types(self):
+        pass
+
+    def _group_parsers_by_log_type(self):
+        grouped_parsers = defaultdict(list)
+        for parser in self._parsers.values():
+            grouped_parsers[parser.log_type].append(parser)
+        return grouped_parsers
+
+    def add_rule(self, user_rule_intent):
+        created_rule = RegexRuleFactory.create_from_intent(user_rule_intent)
+        self._save_rule_definition(created_rule.serialize())
+        created_parsers = created_rule.get_new_parsers(self._parsers)
+        self._save_parsers_definition(parser.serialize() for parser in created_parsers)
+        self._rules.append(created_rule)
+        for parser in created_parsers:
+            self._parsers[parser.name] = parser
+
+    def add_log_type(self, log_type):
+        # TODO Can assume that exists only one LogType object for one log type name
+        pass
+
+    @abstractmethod
+    def _save_rule_definition(self, rule_definition):
+        pass
+
+    @abstractmethod
+    def _save_parsers_definition(self, parser_definitions):
+        pass
+
+    def create_investigation_plan(self, front_input, log_type):
+        matching_parsers, effect_params = self._find_matching_parsers(
+            front_input.line_content, log_type.name
+        )
+        suspected_rules = self._filter_rule_set(matching_parsers)
+        concatenated_parsers = self._create_concatenated_parsers_for_investigation(suspected_rules)
+        #TODO: creating clues base on effect_params
+        #TODO: remove mocks
+        effect_time = datetime(2015, 12, 3, 12, 8, 9)
+        line_source = LineSource('localhost', 'node_1.log', 40)
+        effect_clues = {'effect': Clue((effect_time,), 'node_1.log', line_source)}
+        steps = self._create_steps_in_investigation(
+            concatenated_parsers, suspected_rules, effect_clues
+        )
+        return InvestigationPlan(suspected_rules, steps, effect_clues)
+
+    def get_log_type(self, front_input):
+        # TODO: remove mock
+        matcher = RegexFilenameMatcher('localhost', 'node_1.log', 'default')
+        return LogType('default', [matcher])
+
+    def _find_matching_parsers(self, effect_line_content, log_type_name):
+        """
+        This method finding all parsers from Config base which matching with effect_line_content
+        """
+        matching_parsers = []
+        extracted_params = {}
+        for parser in self._parsers_grouped_by_log_type[log_type_name]:
+            params = parser.get_regex_params(effect_line_content)
+            if params is not None:
+                extracted_params[parser.name] = params
+                matching_parsers.append(parser)
+        return matching_parsers, extracted_params
+
+    def _filter_rule_set(self, parsers_list):
+        """
+        This method finding all rules from Config base which can be fulfilled in
```

Comment:
It is entire rules, parsers and log_types which belongs to single Config object.